### PR TITLE
[Bugfix] Fix Dtype Parsing

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
## Purpose
Currently, the data type is not parsed correctly when a non-string override is provided and we create the default diffusion config. This is because we normalize to a string [here](https://github.com/vllm-project/vllm-omni/blob/a42b7487b0681afab4a5a4b305d7f59c0b158de9/vllm_omni/entrypoints/omni.py#L177) without splitting off the `torch.` prefix, and later look up the torch dtype [here](https://github.com/vllm-project/vllm-omni/blob/a42b7487b0681afab4a5a4b305d7f59c0b158de9/vllm_omni/diffusion/data.py#L447), which causes it to fall back to the default.

E.g.,
```python 
from vllm_omni.entrypoints.omni import Omni
import torch

omni = Omni(model="Tongyi-MAI/Z-Image-Turbo", dtype=torch.float16)
```

-> `[Stage-0] WARNING 02-16 16:08:08 [data.py:463] Unknown dtype string 'torch.float16', defaulting to bfloat16`

This PR fixes the str parsing, and also throws a `TypeError` if a dtype that isn't a str/torch dtype is provided; can be verified by ensuring the dtype is parsed correctly and that we don't see this warning.

### CODE REVIEW

File: vllm_omni/entrypoints/omni.py [hunk 1]
```diff
@@ -173,8 +174,11 @@ def _create_default_diffusion_stage_cfg(self, kwargs: dict[str, Any]) -> dict[st
         # We temporally create a default config for diffusion stage.
         # In the future, we should merge the default config with the user-provided config.
         # TODO: hack, convert dtype to string to avoid non-premitive omegaconf create error.
-        if "dtype" in kwargs:
-            kwargs["dtype"] = str(kwargs["dtype"])
+        if "dtype" in kwargs and not isinstance(kwargs["dtype"], str):
+            if not isinstance(kwargs["dtype"], torch.dtype):
+                raise TypeError("Provided dtype must be a string or torch.dtype")
+            kwargs["dtype"] = str(kwargs["dtype"]).split(".")[-1]
```
Comment (gcanlin) [inline vllm_omni/entrypoints/omni.py]: Maybe consider using `removeprefix` instead of split(".")[-1] for clarity:

```suggestion
            kwargs["dtype"] = str(kwargs["dtype"]).removeprefix("torch.")
```

File: vllm_omni/entrypoints/omni.py [hunk 2]
```diff
@@ -173,8 +174,11 @@ def _create_default_diffusion_stage_cfg(self, kwargs: dict[str, Any]) -> dict[st
         # We temporally create a default config for diffusion stage.
         # In the future, we should merge the default config with the user-provided config.
         # TODO: hack, convert dtype to string to avoid non-premitive omegaconf create error.
-        if "dtype" in kwargs:
-            kwargs["dtype"] = str(kwargs["dtype"])
+        if "dtype" in kwargs and not isinstance(kwargs["dtype"], str):
+            if not isinstance(kwargs["dtype"], torch.dtype):
+                raise TypeError("Provided dtype must be a string or torch.dtype")
+            kwargs["dtype"] = str(kwargs["dtype"]).removeprefix("torch.")
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/omni.py:223]: Error message would be more useful if it included what was passed:
```python
raise TypeError(
    f"Provided dtype must be a string or torch.dtype, "
    f"got {type(kwargs['dtype']).__name__}"
)
```

File: vllm_omni/entrypoints/omni.py [hunk 3]
```diff
@@ -173,8 +174,11 @@ def _create_default_diffusion_stage_cfg(self, kwargs: dict[str, Any]) -> dict[st
         # We temporally create a default config for diffusion stage.
         # In the future, we should merge the default config with the user-provided config.
         # TODO: hack, convert dtype to string to avoid non-premitive omegaconf create error.
-        if "dtype" in kwargs:
-            kwargs["dtype"] = str(kwargs["dtype"])
+        if "dtype" in kwargs and not isinstance(kwargs["dtype"], str):
+            if not isinstance(kwargs["dtype"], torch.dtype):
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/omni.py:221]: Worth a quick test — `torch.float16`, `"float16"`, and something invalid like `int` — to prevent regressions.

**Review by lishunyang12 — COMMENTED** [review]
LGTM. Two small suggestions.

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer

File: vllm_omni/entrypoints/omni.py (hunk 3 above)
Comment (alex-jw-brooks) [reply vllm_omni/entrypoints/omni.py:221]: Sure - I added some happy / sad path tests in the omni diffusion tests that would have caught this. When this is eventually refactored for some of the existing todos, we can also make those tests more atomic

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer

File: vllm_omni/entrypoints/omni.py (hunk 3 above)
Comment (tzhouam) [reply vllm_omni/entrypoints/omni.py:221]: Should the test be place in the entrypoints instead of omni diffusions?

Comment (alex-jw-brooks) [reply vllm_omni/entrypoints/omni.py:221]: @tzhouam I think it's a reasonable place to put it for now, mostly just because of the test using similar monkeypatching and this code path not being very easy to test in isolation at the moment - happy to move it if you think there is a better place for it though

**Review by lishunyang12 — COMMENTED** [review]
Left one comment on the tests.

File: tests/entrypoints/test_omni_diffusion.py [hunk 4]
```diff
@@ -1340,3 +1341,95 @@ def _fake_loader(
...
+    import vllm_omni.entrypoints.omni as omni_module
+    from vllm_omni.entrypoints.omni import Omni
+
+    monkeypatch.setattr(
+        "vllm_omni.entrypoints.utils.load_stage_configs_from_model",
+        _fake_loader,
+        raising=False,
+    )
+
+    monkeypatch.setattr(omni_module, "load_stage_configs_from_model", _fake_loader)
```
Comment (lishunyang12) [inline tests/entrypoints/test_omni_diffusion.py]: This is a bare comparison, not an assertion -- silently passes even if `stage_type` is wrong.

```suggestion
    assert omni.stage_configs[0]["stage_type"] == "diffusion"
```

File: vllm_omni/entrypoints/omni.py (hunk 3 above)
Comment (lishunyang12) [reply vllm_omni/entrypoints/omni.py:221]: Looks good, thanks for adding them.

File: tests/entrypoints/test_omni_diffusion.py (hunk 4 above)
Comment (alex-jw-brooks) [reply tests/entrypoints/test_omni_diffusion.py]: Good catch, thanks - It looks like I didn't catch some changes in the test structure when I rebased also, should be fixed now

**Comment by tzhouam** [comment]
@fake0fan Could you review this one?

**Comment by fake0fan** [comment]
This code modification is not complicated, and others have already offered many suggestions, so I have no further comments.

**Review by gcanlin — APPROVED** [review]
